core/src/main/scala/zio/logging/LogAppender.scala

```diff
@@ -0,0 +1,80 @@
+package zio.logging
+
+import zio.{ UIO, URIO, ZIO }
+import zio.console._
+import zio.ZLayer
+import izumi.reflect.Tag
+import zio.ZManaged
+import java.nio.file.Files
+import java.nio.file.Paths
+import zio.Has
+
+/**
+ * Represets log writer function that turns A into String and put in console or save to file.
+ */
+object LogAppender {
+
+  trait Service[A] { self =>
+
+    def write(ctx: LogContext, msg: => A): UIO[Unit]
+
+    final def filter(fn: (LogContext, => A) => Boolean): Service[A] =
+      new Service[A] {
+        override def write(ctx: LogContext, msg: => A): zio.UIO[Unit] =
+          if (fn(ctx, msg))
+            self.write(ctx, msg)
+          else
+            ZIO.unit
+      }
+  }
+
+  def make[R, A: Tag](
+    format0: LogFormat[A],
+    write0: (LogContext, => String) => URIO[R, Unit]
+  ): ZLayer[R, Nothing, Appender[A]] =
+    ZIO
+      .access[R](env =>
+        new Service[A] {
+
+          override def write(ctx: LogContext, msg: => A): UIO[Unit] =
+            write0(ctx, format0.format(ctx, msg)).provide(env)
+        }
+      )
+      .toLayer
+
+  def console[A: Tag](logLevel: LogLevel, format: LogFormat[A]): ZLayer[Console, Nothing, Appender[A]] =
+    make[Console, A](format, (_, line) => putStrLn(line)).map(appender =>
+      Has(appender.get.filter((ctx, _) => ctx.get(LogAnnotation.Level) >= logLevel))
+    )
+
+  def consoleErr[A: Tag](logLevel: LogLevel, format: LogFormat[A]): ZLayer[Console, Nothing, Appender[A]] =
+    make[Console, A](
+      format,
+      (ctx, msg) =>
+        if (ctx.get(LogAnnotation.Level) == LogLevel.Error)
+          putStrLnErr(msg)
+        else
+          putStrLn(msg)
+    ).map(appender => Has(appender.get.filter((ctx, _) => ctx.get(LogAnnotation.Level) >= logLevel)))
+
+  def file[A: Tag](filename: String, format0: LogFormat[A]) =
+    ZManaged.makeEffect {
+      val path = Paths.get(filename)
+      Files.newBufferedWriter(path)
+    }(_.close())
+      .map(writer =>
+        new Service[A] {
+
+          override def write(ctx: LogContext, msg: => A): UIO[Unit] =
+            ZIO.effectTotal(writer.write(format0.format(ctx, msg)))
```

Comment:
Why `.effectTotal` if file writer can fail with some IO exception?

Reply:
it can indeed. question is what you can do about this? technically when you writing to console or to slf4j you may get exception too. 

that will start discussion if we should include errors in logging interface. we decided to not for the time being.

Reply:
Ok, but if `writer.write` fails the whole fiber will die, correct?

Reply:
we can ignore it but I don't really like sweeping stuff under rug approach. so to answer you question yes it's desired i want program to exit if logging module failed to write log to file.

would you expect different behavior?

Reply:
Yes. I'd expect the library to be safe to use so I don't have to try/catch library calls. We're behind exception driven development. If some library API can fail for some reason, I'd like to know it.

Reply:
for me this will very uncomfortable api that I need to handle error from logging that I can do nothing about. Try to visual code with this error and how would you handle this.